src/tag.c

```diff
@@ -0,0 +1,112 @@
+#include "utils.h"
+
+SEXP have_name(SEXP x) {
+  SEXP nms = PROTECT(Rf_getAttrib(x, R_NamesSymbol));
+  R_xlen_t n = Rf_xlength(x);
+  SEXP out = PROTECT(Rf_allocVector(LGLSXP, n));
+
+  if (nms == R_NilValue) {
+    for (R_xlen_t i = 0; i < n; ++i) {
+      SET_LOGICAL_ELT(out, i, 0);
+    }
+  } else {
+    for (R_xlen_t i = 0; i < n; ++i) {
+      SEXP nm_i = STRING_ELT(nms, i);
+      // TODO cache empty string
+      SET_LOGICAL_ELT(out, i, nm_i != NA_STRING & nm_i != chr_empty);
+    }
+  }
+
+  UNPROTECT(2);
+  return out;
+}
+
+SEXP new_tag(SEXP tagName, SEXP varArgs, SEXP noWS, SEXP renderHook) {
+  R_xlen_t n = Rf_xlength(varArgs);
+
+  // TODO validate that varArgs is a list
```

Comment:
It is documented that `varArgs` must be a list but there is a test where this is not the case. Should we allow `varArgs` to be e.g. a character and simply wrap it in a list?